New blog post about Che deployment on AKS (#29)
eclipse-che/blog: _posts/2022-07-25-installing-eclipse-che-on-aks.adoc

NOTE: The field `identityToken` is specific to {prod} installations made for Kubernetes only and ignored for OpenShift.

== Challenge 2: No way to obtain proper access token accepted by AKS
This is a continuation of Challenge 1 in a way that after we taught {prod} to leverage `access_token` instead of `id_token` to access AKS, we need to make sure this `access_token` is accepted by AKS.

When AKS integration with AAD is enabled, AKS expects OpenID Connect (OIDC) tokens issued to the certain application, pre-registered in AAD: 'Azure Kubernetes Service AAD Server'.

.Azure Kubernetes Service AAD Server application
image::/assets/img/installing-eclipse-che-on-aks/aks-aad-server-app.png[Azure Kubernetes Service AAD Server]

If we leave default settings of OAuth2 Proxy, leveraged in Gateway, it uses standard set of OIDC scopes and for them AAD returns an `access_token` that can be used against Microsoft Graph, not AKS.

In order to obtain a token suitable for AKS, additional scope should be configured for OAuth2 Proxy, representing Azure Kubernetes Service AAD Server application:
[source,yaml]
----
6dae42f8-4368-4678-94ff-3960e28e3630/user.read
----

> where GUID is a Application ID for AKS registered in AAD.

Reviewer: ```suggestion
where the ID is the Application ID of the AKS instance registered in AAD.
```

[source,yaml]
----
spec:
  networking:
    auth:
      identityProviderURL: https://sts.windows.net/{TENANT_ID}/v2.0/
      identityToken: access_token
      oAuthClientName: {CLIENT_ID}
      oAuthSecret: {CLIENT_SECRET}
      oAuthScope: openid email profile 6dae42f8-4368-4678-94ff-3960e28e3630/user.read
  components:
    cheServer:
      extraProperties:
        CHE_OIDC_AUTH__SERVER__URL: https://sts.windows.net/{TENANT_ID}/v2.0/
        CHE_OIDC_EMAIL__CLAIM: unique_name 
----

* `TENANT_ID` - Directory (tenant) ID, see Figure 3.
* `CLIENT_ID` - Application (client) ID, see Figure 3.
* `CLIENT_SECRET` - Client secret, you can manage it in 'Certificates & secret' section

.Registered Che application
image::/assets/img/installing-eclipse-che-on-aks/azure-che-demo-app.png[Registered Che application]

WARNING: Don't forget to configure API permissions to authorize your application to call AKS Server API.

.AKS API permissions
image::/assets/img/installing-eclipse-che-on-aks/aks-api-permissions.png[AKS API permissions]

> After all preparations are completed, `chectl` deploy command can be used to install {prod} on AKS using yaml file above:

Reviewer: ```suggestion
After the {prod} App configuration in Azure is completed, the command `chectl server:deploy` can be used to install {prod} on AKS using the `YAML` file above:
```

[source,shell]
----
chectl server:deploy \
       --platform=k8s \
       --installer=operator \
       --che-operator-cr-patch-yaml=che.yaml \
       --skip-oidc-provider-check \
       --skip-cert-manager \
       --domain=eclipse-che-demo.mydomain.com
----

> NOTE: In our case we already configured `cert-manager` and created `domain` according to the link:https://www.eclipse.org/che/docs/che-7/installation-guide/installing-che-on-microsoft-azure/[instruction].

Reviewer: ```suggestion
NOTE: In our case we already configured `cert-manager` and created `domain` according to the link:https://www.eclipse.org/che/docs/che-7/installation-guide/installing-che-on-microsoft-azure/[old (7.41) instructions].
```
